## Problem 11: Largest Product in a Grid

In the 20×20 grid below, four numbers along a diagonal line have been marked in red.

\begin{align}
08\ 02\ 22\ 97\ 38\ 15\ 00\ 40\ 00\ 75\ 04\ 05\ 07\ 78\ 52\ 12\ 50\ 77\ 91\ 08 \\ 
49\ 49\ 99\ 40\ 17\ 81\ 18\ 57\ 60\ 87\ 17\ 40\ 98\ 43\ 69\ 48\ 04\ 56\ 62\ 00 \\ 
81\ 49\ 31\ 73\ 55\ 79\ 14\ 29\ 93\ 71\ 40\ 67\ 53\ 88\ 30\ 03\ 49\ 13\ 36\ 65 \\ 
52\ 70\ 95\ 23\ 04\ 60\ 11\ 42\ 69\ 24\ 68\ 56\ 01\ 32\ 56\ 71\ 37\ 02\ 36\ 91 \\ 
22\ 31\ 16\ 71\ 51\ 67\ 63\ 89\ 41\ 92\ 36\ 54\ 22\ 40\ 40\ 28\ 66\ 33\ 13\ 80 \\ 
24\ 47\ 32\ 60\ 99\ 03\ 45\ 02\ 44\ 75\ 33\ 53\ 78\ 36\ 84\ 20\ 35\ 17\ 12\ 50 \\ 
32\ 98\ 81\ 28\ 64\ 23\ 67\ 10\ {\color{red}2\color{red}6}\ 38\ 40\ 67\ 59\ 54\ 70\ 66\ 18\ 38\ 64\ 70 \\ 
67\ 26\ 20\ 68\ 02\ 62\ 12\ 20\ 95\ {\color{red}6\color{red}3}\ 94\ 39\ 63\ 08\ 40\ 91\ 66\ 49\ 94\ 21 \\ 
24\ 55\ 58\ 05\ 66\ 73\ 99\ 26\ 97\ 17\ {\color{red}7\color{red}8}\ 78\ 96\ 83\ 14\ 88\ 34\ 89\ 63\ 72 \\ 
21\ 36\ 23\ 09\ 75\ 00\ 76\ 44\ 20\ 45\ 35\ {\color{red}1\color{red}4}\ 00\ 61\ 33\ 97\ 34\ 31\ 33\ 95 \\ 
78\ 17\ 53\ 28\ 22\ 75\ 31\ 67\ 15\ 94\ 03\ 80\ 04\ 62\ 16\ 14\ 09\ 53\ 56\ 92 \\ 
16\ 39\ 05\ 42\ 96\ 35\ 31\ 47\ 55\ 58\ 88\ 24\ 00\ 17\ 54\ 24\ 36\ 29\ 85\ 57 \\ 
86\ 56\ 00\ 48\ 35\ 71\ 89\ 07\ 05\ 44\ 44\ 37\ 44\ 60\ 21\ 58\ 51\ 54\ 17\ 58 \\ 
19\ 80\ 81\ 68\ 05\ 94\ 47\ 69\ 28\ 73\ 92\ 13\ 86\ 52\ 17\ 77\ 04\ 89\ 55\ 40 \\ 
04\ 52\ 08\ 83\ 97\ 35\ 99\ 16\ 07\ 97\ 57\ 32\ 16\ 26\ 26\ 79\ 33\ 27\ 98\ 66 \\ 
88\ 36\ 68\ 87\ 57\ 62\ 20\ 72\ 03\ 46\ 33\ 67\ 46\ 55\ 12\ 32\ 63\ 93\ 53\ 69 \\ 
04\ 42\ 16\ 73\ 38\ 25\ 39\ 11\ 24\ 94\ 72\ 18\ 08\ 46\ 29\ 32\ 40\ 62\ 76\ 36 \\ 
20\ 69\ 36\ 41\ 72\ 30\ 23\ 88\ 34\ 62\ 99\ 69\ 82\ 67\ 59\ 85\ 74\ 04\ 36\ 16 \\ 
20\ 73\ 35\ 29\ 78\ 31\ 90\ 01\ 74\ 31\ 49\ 71\ 48\ 86\ 81\ 16\ 23\ 57\ 05\ 54 \\ 
01\ 70\ 54\ 71\ 83\ 51\ 54\ 69\ 16\ 92\ 33\ 48\ 61\ 43\ 52\ 01\ 89\ 19\ 67\ 48 \\ 
\end{align}

The product of these numbers is 26 × 63 × 78 × 14 = 1,788,696.

What is the greatest product of four adjacent numbers in the same direction (up, down, left, right, or diagonally) in the 20×20 grid?

In [1]:
number_grid = [
    [ 8, 2,22,97,38,15, 0,40, 0,75, 4, 5, 7,78,52,12,50,77,91, 8],
    [49,49,99,40,17,81,18,57,60,87,17,40,98,43,69,48, 4,56,62, 0],
    [81,49,31,73,55,79,14,29,93,71,40,67,53,88,30, 3,49,13,36,65],
    [52,70,95,23, 4,60,11,42,69,24,68,56, 1,32,56,71,37, 2,36,91],
    [22,31,16,71,51,67,63,89,41,92,36,54,22,40,40,28,66,33,13,80],
    [24,47,32,60,99, 3,45, 2,44,75,33,53,78,36,84,20,35,17,12,50],
    [32,98,81,28,64,23,67,10,26,38,40,67,59,54,70,66,18,38,64,70],
    [67,26,20,68, 2,62,12,20,95,63,94,39,63, 8,40,91,66,49,94,21],
    [24,55,58, 5,66,73,99,26,97,17,78,78,96,83,14,88,34,89,63,72],
    [21,36,23, 9,75, 0,76,44,20,45,35,14, 0,61,33,97,34,31,33,95],
    [78,17,53,28,22,75,31,67,15,94, 3,80, 4,62,16,14, 9,53,56,92],
    [16,39, 5,42,96,35,31,47,55,58,88,24, 0,17,54,24,36,29,85,57],
    [86,56, 0,48,35,71,89, 7, 5,44,44,37,44,60,21,58,51,54,17,58],
    [19,80,81,68, 5,94,47,69,28,73,92,13,86,52,17,77, 4,89,55,40],
    [ 4,52, 8,83,97,35,99,16, 7,97,57,32,16,26,26,79,33,27,98,66],
    [88,36,68,87,57,62,20,72, 3,46,33,67,46,55,12,32,63,93,53,69],
    [ 4,42,16,73,38,25,39,11,24,94,72,18, 8,46,29,32,40,62,76,36],
    [20,69,36,41,72,30,23,88,34,62,99,69,82,67,59,85,74, 4,36,16],
    [20,73,35,29,78,31,90, 1,74,31,49,71,48,86,81,16,23,57, 5,54],
    [ 1,70,54,71,83,51,54,69,16,92,33,48,61,43,52, 1,89,19,67,48],
]

In [2]:
# Only need to search down, right, diagonal(down and right), and diagonal(down and left).
# Searching up is is the same as down, left the same as right and diagonal(up and l/r) the same as diagonal(down and r/l)


''' 
Takes a start point in the x and y directions, the direction to move in x and y, and the number of steps to make.
It then calculates the product of the numbers in the array that are found when moving in the way defined by the inputs.

eg:
>>> grid_multiplication(0, 0, 1, 1, 4)
>>> 279496
'''

def grid_multiplication(start_x : int, start_y : int, step_x : int, step_y : int, n_steps : int) -> int:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Takes a start point in the x and y directions, the direction to move in x and y, and the number of steps to make.
        It then calculates the product of the numbers in the array that are found when moving in the way defined by the inputs.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : start_x : The x index to start at.
        : int : start_y : The y index to start at.
        : int : step_x : The distance to step in the x direction. Not absolute, so a step_x of -1 will step backwards.
        : int : step_y : The distance to step in the y direction. Not absolute, so a step_x of -1 will step backwards.
        : int : n_steps : The number of steps in both directions.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : int : total_product : The total product of the numbers in the grid found by the movement defined by the inputs.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> grid_multiplication(0, 0, 1, 1, 4)
        >>> 279496
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(start_x, int): raise ValueError('Please enter an integer > 0 for the start_x argument.')
    if start_x < 0: raise ValueError('Please enter an integer >= 0 for the start_x argument.')
    if not isinstance(start_y, int): raise ValueError('Please enter an integer > 0 for the start_y argument.')
    if start_y < 0: raise ValueError('Please enter an integer >= 0 for the start_y argument.')
    if not isinstance(step_x, int): raise ValueError('Please enter an integer > 0 for the step_x argument.')
    if not isinstance(step_y, int): raise ValueError('Please enter an integer > 0 for the step_y argument.')
    if not isinstance(n_steps, int): raise ValueError('Please enter an integer > 0 for the n_steps argument.')
    if n_steps < 0: raise ValueError('Please enter an integer >= 0 for the n_steps argument.')
    
    total_product = 1
    
    # Step through the grid indices as defined by the function inputs.
    for step in range(n_steps):
        x = start_x + (step * step_x)
        y = start_y + (step * step_y)
        
        # Check x and y are valid
        if y > len(number_grid) - 1:
            return -1
        if x > len(number_grid[y]) - 1:
            return -1
        
        # Find the number in the grid
        number_in_grid = number_grid[y][x]
        
        # Multiply it to the total product.
        total_product *= number_in_grid
    
    return total_product

In [3]:
''' 
Returns the numbers used in the grid multiplication (for use in stating the solution to the problem).

eg:
>>> grid_retrieval(0, 0, 1, 1, 4)
>>> ['8', '49', '31', '23']
'''

def grid_retrieval(start_x : int, start_y : int, step_x : int, step_y : int, n_steps : int) -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Takes a start point in the x and y directions, the direction to move in x and y, and the number of steps to make.
        It then calculates the product of the numbers in the array that are found when moving in the way defined by the inputs.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : start_x : The x index to start at.
        : int : start_y : The y index to start at.
        : int : step_x : The distance to step in the x direction. Not absolute, so a step_x of -1 will step backwards.
        : int : step_y : The distance to step in the y direction. Not absolute, so a step_x of -1 will step backwards.
        : int : n_steps : The number of steps in both directions.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : numbers_in_grid : The numbers in the grid used in the grid multiplication.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> grid_retrieval(0, 0, 1, 1, 4)
        >>> ['8', '49', '31', '23']
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(start_x, int): raise ValueError('Please enter an integer > 0 for the start_x argument.')
    if start_x < 0: raise ValueError('Please enter an integer >= 0 for the start_x argument.')
    if not isinstance(start_y, int): raise ValueError('Please enter an integer > 0 for the start_y argument.')
    if start_y < 0: raise ValueError('Please enter an integer >= 0 for the start_y argument.')
    if not isinstance(step_x, int): raise ValueError('Please enter an integer > 0 for the step_x argument.')
    if not isinstance(step_y, int): raise ValueError('Please enter an integer > 0 for the step_y argument.')
    if not isinstance(n_steps, int): raise ValueError('Please enter an integer > 0 for the n_steps argument.')
    if n_steps < 0: raise ValueError('Please enter an integer >= 0 for the n_steps argument.')
    
    numbers_in_grid = []
    
    # Step through the grid indices as defined by the function inputs.
    for step in range(n_steps):
        x = start_x + (step * step_x)
        y = start_y + (step * step_y)
        
        # Check x and y are valid
        if y > len(number_grid) - 1:
            return []
        if x > len(number_grid[y]) - 1:
            return []
        
        # Find the number in the grid
        number_in_grid = number_grid[y][x]
        
        # Multiply it to the total product.
        numbers_in_grid.append(str(number_in_grid))
    
    return numbers_in_grid

In [4]:
''' 
Loops through each position in the number grid and tests if it can search each of the four directions.
If it can, it performs the grid multiplication and compares it to the current greatest product, if it is higher
then the new greatest product is stored as well as additional information.

eg:
>>> grid_search(3)
>>> (811502, 6, 12, 'diagonal(down and left)')
'''

def grid_search(search_distance : int) -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Loops through each position in the number grid and tests if it can search each of the four directions.
        If it can, it performs the grid multiplication and compares it to the current greatest product, if it is higher
        then the new greatest product is stored as well as additional information.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : search_distance : The length of the product in the grid.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : Contains the following information:
            : int : greatest_product : The largest product found in the grid.
            : int : greatest_product_x : The starting x index for the product.
            : int : greatest_product_y : The starting y index for the product.
            : str : greatest_product_direction : The direction from the starting x and y to move.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> grid_search(3)
        >>> (811502, 6, 12, 'diagonal(down and left)')
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(search_distance, int): raise ValueError('Please enter an integer > 0 for the search_distance argument.')
    if search_distance < 0: raise ValueError('Please enter an integer >= 0 for the search_distance argument.')
    
    greatest_product = -1
    greatest_product_x = -1
    greatest_product_y = -1 
    greatest_product_direction = ''
    
    for current_y in range(len(number_grid)):
        
        for current_x in range(len(number_grid[current_y])):
            
            # Search down
            if (current_y + search_distance < len(number_grid)):
                test_product = grid_multiplication(current_x, current_y, 0, 1, search_distance)
                if test_product > greatest_product: 
                    greatest_product = test_product
                    greatest_product_x = current_x
                    greatest_product_y = current_y
                    greatest_product_direction = 'down'
            
            # Search right
            if (current_x + search_distance < len(number_grid[current_y])):
                test_product = grid_multiplication(current_x, current_y, 1, 0, search_distance)
                if test_product > greatest_product: 
                    greatest_product = test_product
                    greatest_product_x = current_x
                    greatest_product_y = current_y
                    greatest_product_direction = 'right'
            
            # Search diagonal(down and right)
            if (current_y + search_distance < len(number_grid)) and (current_x + search_distance < len(number_grid[current_y])):
                test_product = grid_multiplication(current_x, current_y, 1, 1, search_distance)
                if test_product > greatest_product: 
                    greatest_product = test_product
                    greatest_product_x = current_x
                    greatest_product_y = current_y
                    greatest_product_direction = 'diagonal(down and right)'
            
            # Search diagonal(down and left)
            if (current_y + search_distance < len(number_grid)) and (current_x - search_distance >= -1):
                test_product = grid_multiplication(current_x, current_y, -1, 1, search_distance)
                if test_product > greatest_product: 
                    greatest_product = test_product
                    greatest_product_x = current_x
                    greatest_product_y = current_y
                    greatest_product_direction = 'diagonal(down and left)'
            
    return greatest_product, greatest_product_x, greatest_product_y, greatest_product_direction

In [5]:
product_chain_length = 4

solution = grid_search(product_chain_length)

if solution[3] == 'down':
    solution_numbers = grid_retrieval(solution[1], solution[2], 0, 1, product_chain_length)
if solution[3] == 'right':
    solution_numbers = grid_retrieval(solution[1], solution[2], 1, 0, product_chain_length)
if solution[3] == 'diagonal(down and right)':
    solution_numbers = grid_retrieval(solution[1], solution[2], 1, 1, product_chain_length)
if solution[3] == 'diagonal(down and left)':
    solution_numbers = grid_retrieval(solution[1], solution[2], -1, 1, product_chain_length)

print(f'The greatest product of {product_chain_length} adjacent numbers in the same direction is:')
print(' * '.join(solution_numbers) + f' = {solution[0]:,}')
print(f'which are the {product_chain_length} numbers starting at positon ({solution[1]}, {solution[2]}) and moving {solution[3]}.')

The greatest product of 4 adjacent numbers in the same direction is:
89 * 94 * 97 * 87 = 70,600,674
which are the 4 numbers starting at positon (6, 12) and moving diagonal(down and left).


### Problem 11 Solution:

The greatest product of 4 adjacent numbers in the same direction is:

$89 * 94 * 97 * 87 = 70,600,674$

which are the 4 numbers starting at positon (6, 12) and moving diagonal(down and left). Here are the numbers highlighted in the number grid:

\begin{align}
08\ 02\ 22\ 97\ 38\ 15\ 00\ 40\ 00\ 75\ 04\ 05\ 07\ 78\ 52\ 12\ 50\ 77\ 91\ 08 \\ 
49\ 49\ 99\ 40\ 17\ 81\ 18\ 57\ 60\ 87\ 17\ 40\ 98\ 43\ 69\ 48\ 04\ 56\ 62\ 00 \\ 
81\ 49\ 31\ 73\ 55\ 79\ 14\ 29\ 93\ 71\ 40\ 67\ 53\ 88\ 30\ 03\ 49\ 13\ 36\ 65 \\ 
52\ 70\ 95\ 23\ 04\ 60\ 11\ 42\ 69\ 24\ 68\ 56\ 01\ 32\ 56\ 71\ 37\ 02\ 36\ 91 \\ 
22\ 31\ 16\ 71\ 51\ 67\ 63\ 89\ 41\ 92\ 36\ 54\ 22\ 40\ 40\ 28\ 66\ 33\ 13\ 80 \\ 
24\ 47\ 32\ 60\ 99\ 03\ 45\ 02\ 44\ 75\ 33\ 53\ 78\ 36\ 84\ 20\ 35\ 17\ 12\ 50 \\ 
32\ 98\ 81\ 28\ 64\ 23\ 67\ 10\ 26\ 38\ 40\ 67\ 59\ 54\ 70\ 66\ 18\ 38\ 64\ 70 \\ 
67\ 26\ 20\ 68\ 02\ 62\ 12\ 20\ 95\ 63\ 94\ 39\ 63\ 08\ 40\ 91\ 66\ 49\ 94\ 21 \\ 
24\ 55\ 58\ 05\ 66\ 73\ 99\ 26\ 97\ 17\ 78\ 78\ 96\ 83\ 14\ 88\ 34\ 89\ 63\ 72 \\ 
21\ 36\ 23\ 09\ 75\ 00\ 76\ 44\ 20\ 45\ 35\ 14\ 00\ 61\ 33\ 97\ 34\ 31\ 33\ 95 \\ 
78\ 17\ 53\ 28\ 22\ 75\ 31\ 67\ 15\ 94\ 03\ 80\ 04\ 62\ 16\ 14\ 09\ 53\ 56\ 92 \\ 
16\ 39\ 05\ 42\ 96\ 35\ 31\ 47\ 55\ 58\ 88\ 24\ 00\ 17\ 54\ 24\ 36\ 29\ 85\ 57 \\ 
86\ 56\ 00\ 48\ 35\ 71\ {\color{red}8\color{red}9}\ 07\ 05\ 44\ 44\ 37\ 44\ 60\ 21\ 58\ 51\ 54\ 17\ 58 \\ 
19\ 80\ 81\ 68\ 05\ {\color{red}9\color{red}4}\ 47\ 69\ 28\ 73\ 92\ 13\ 86\ 52\ 17\ 77\ 04\ 89\ 55\ 40 \\ 
04\ 52\ 08\ 83\ {\color{red}9\color{red}7}\ 35\ 99\ 16\ 07\ 97\ 57\ 32\ 16\ 26\ 26\ 79\ 33\ 27\ 98\ 66 \\ 
88\ 36\ 68\ {\color{red}8\color{red}7}\ 57\ 62\ 20\ 72\ 03\ 46\ 33\ 67\ 46\ 55\ 12\ 32\ 63\ 93\ 53\ 69 \\ 
04\ 42\ 16\ 73\ 38\ 25\ 39\ 11\ 24\ 94\ 72\ 18\ 08\ 46\ 29\ 32\ 40\ 62\ 76\ 36 \\ 
20\ 69\ 36\ 41\ 72\ 30\ 23\ 88\ 34\ 62\ 99\ 69\ 82\ 67\ 59\ 85\ 74\ 04\ 36\ 16 \\ 
20\ 73\ 35\ 29\ 78\ 31\ 90\ 01\ 74\ 31\ 49\ 71\ 48\ 86\ 81\ 16\ 23\ 57\ 05\ 54 \\ 
01\ 70\ 54\ 71\ 83\ 51\ 54\ 69\ 16\ 92\ 33\ 48\ 61\ 43\ 52\ 01\ 89\ 19\ 67\ 48 \\ 
\end{align}